local_planner/src/local_planner.cpp

```diff
@@ -397,6 +400,54 @@ void LocalPlanner::getReference() {
     }
   }
 
+  // Unwrap yaw angle for NMPC
+  Eigen::VectorXd eig_unwrapped_yaw_ref = ref_body_plan_.col(5);
+
+  std::vector<double> wrapped_yaw_ref, unwrapped_yaw_ref;
+  quad_utils::eigenToVector(eig_unwrapped_yaw_ref, wrapped_yaw_ref);
+
+  // If first plan, keep nominal yaw, else align initial yaw state with previous
```

Comment:
I think these changes are more complicated than they need to be, we shouldn't need to track prior yaw values or do anything with quotients, to my knowledge.

I think what we want is to set the first element of the reference plan (so eig_unwrapped_yaw_ref[0]) to some value that is less than M_PI from the current yaw. For example, if current yaw = 3 and eig_unwrapped_yaw_ref[0] = -3, we should update with eig_unwrapped_yaw_ref[0]) 2 * M_PI - eig_unwrapped_yaw_ref[0] (the check for this is just abs(diff) >= M_PI); this ensures that there isn't a false difference between the current state and the reference.

Once that value has been updated we use the unwrap function to get the rest of the unwrapped reference. Since that function just iterates through the vector and checks for differences of > M_PI, it should be able to update the rest of the reference to be continuous (although it may result in abs(yaw) > M_PI so we'll need to make sure the bounds on yaw in IPOPT are more than that).

As an aside, it looks like math_utils::unwrap expects an std::vector but that can be updated to be templated so we can just use the Eigen vector directly:

```
template <typename VectorType>
VectorType unwrap(const VectorType& data) {
  VectorType data_unwrapped = data;
  ...
  return data_unwrapped;
}
```

Reply:
Sorry if this thread is a bit outdated, just now returning to look at things.

I'm not quite following your math there - a yaw of -361 deg would be unwrapped to -1 deg, not 179. Also quaternion to roll/pitch/yaw is in fact a 1-1 mapping so long as the system is not at a singularity (pitch = +- pi/2) (see [link](https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles). Rewriting NMPC with quaternions is really only needed if you need to plan through that singularity (e.g. doing a backflip).

All this is to say that correctly unwrapping the reference is 1) possible, and 2) the simplest and most robust way to solve the problem, since it does not require keeping track of any histories.

In case my above post wasn't clear, "correctly unwrapping" the reference means taking the reference which is in the domain (-pi, pi] as you say, and performing the unwrapping method which would give a continuous reference in the domain (-inf, inf) and therefore can rotate any number of times. That unwrapped reference is then passed into the NLP.

Reply:
Hey, 


Sorry for these long updates as well. It's getting very busy as I plan to finish up some of the work that I am doing for my master's thesis and I don't think I can dive deeper into the code without some effort. I think we are agreeing on the same thing, that as long as the reference is in the domain of $(-\pi, \pi]$, this current method should be fine. 

However, I don't think it extends to anything beyond that. I would like to think of some elegant solution to extending it such that the reference trajectory can be a larger domain without rehaul of some of the code (and I thought that the solution I proposed would have solved that), but I don't think I'll be able to get to this anytime soon. I fixed some minor bug of ensuring that the current state is unwrapped as well. However, I still get the following behavior:
- When yaw reference trajectory goes past. the current yaw is stuck at $\-\pi$ and won't go further. See video. And see image of the current state output

https://github.com/robomechanics/quad-sdk/assets/85125684/66b0e5c8-f428-4414-bb4e-0712a0b861bb

![quad_sdk_unwrapped_nmpc](https://github.com/robomechanics/quad-sdk/assets/85125684/d9f543b1-d3c1-445e-8d6e-e0613d32c11e)



I am not sure if that behavior is due to how the MPC is formulated, but I am not sure what to make of this behavior. Let me know if I might be missing something.

Additionally, I have kept the extra "quotient" code for now, but I can get rid of it to clean up the code if these are the behaviors that we want. I don't think they are impacting anything as of right now, but thought to keep it in case I found some solution.